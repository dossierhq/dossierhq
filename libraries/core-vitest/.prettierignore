/.rush/
/lib/
/temp/
/tsconfig.json
/CHANGELOG.md
