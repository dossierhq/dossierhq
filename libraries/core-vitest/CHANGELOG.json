{
  "name": "@dossierhq/core-vitest",
  "entries": [
    {
      "version": "0.2.17",
      "tag": "@dossierhq/core-vitest_v0.2.17",
      "date": "Tue, 13 Jun 2023 15:58:39 GMT",
      "comments": {}
    },
    {
      "version": "0.2.16",
      "tag": "@dossierhq/core-vitest_v0.2.16",
      "date": "Tue, 09 May 2023 22:47:11 GMT",
      "comments": {}
    },
    {
      "version": "0.2.15",
      "tag": "@dossierhq/core-vitest_v0.2.15",
      "date": "Tue, 09 May 2023 22:20:37 GMT",
      "comments": {}
    },
    {
      "version": "0.2.14",
      "tag": "@dossierhq/core-vitest_v0.2.14",
      "date": "Sun, 23 Apr 2023 21:31:27 GMT",
      "comments": {}
    },
    {
      "version": "0.2.13",
      "tag": "@dossierhq/core-vitest_v0.2.13",
      "date": "Tue, 11 Apr 2023 19:08:01 GMT",
      "comments": {}
    },
    {
      "version": "0.2.12",
      "tag": "@dossierhq/core-vitest_v0.2.12",
      "date": "Tue, 11 Apr 2023 18:39:14 GMT",
      "comments": {}
    },
    {
      "version": "0.2.11",
      "tag": "@dossierhq/core-vitest_v0.2.11",
      "date": "Tue, 11 Apr 2023 15:53:50 GMT",
      "comments": {}
    },
    {
      "version": "0.2.10",
      "tag": "@dossierhq/core-vitest_v0.2.10",
      "date": "Tue, 04 Apr 2023 09:26:20 GMT",
      "comments": {}
    },
    {
      "version": "0.2.9",
      "tag": "@dossierhq/core-vitest_v0.2.9",
      "date": "Mon, 27 Mar 2023 13:49:59 GMT",
      "comments": {}
    },
    {
      "version": "0.2.8",
      "tag": "@dossierhq/core-vitest_v0.2.8",
      "date": "Mon, 27 Mar 2023 11:41:48 GMT",
      "comments": {}
    },
    {
      "version": "0.2.7",
      "tag": "@dossierhq/core-vitest_v0.2.7",
      "date": "Wed, 22 Mar 2023 21:32:23 GMT",
      "comments": {}
    },
    {
      "version": "0.2.6",
      "tag": "@dossierhq/core-vitest_v0.2.6",
      "date": "Wed, 22 Mar 2023 21:17:29 GMT",
      "comments": {}
    },
    {
      "version": "0.2.5",
      "tag": "@dossierhq/core-vitest_v0.2.5",
      "date": "Tue, 14 Mar 2023 17:47:19 GMT",
      "comments": {}
    },
    {
      "version": "0.2.4",
      "tag": "@dossierhq/core-vitest_v0.2.4",
      "date": "Tue, 07 Mar 2023 21:38:54 GMT",
      "comments": {}
    },
    {
      "version": "0.2.3",
      "tag": "@dossierhq/core-vitest_v0.2.3",
      "date": "Sat, 18 Feb 2023 11:05:21 GMT",
      "comments": {}
    },
    {
      "version": "0.2.2",
      "tag": "@dossierhq/core-vitest_v0.2.2",
      "date": "Tue, 14 Feb 2023 18:18:06 GMT",
      "comments": {}
    },
    {
      "version": "0.2.1",
      "tag": "@dossierhq/core-vitest_v0.2.1",
      "date": "Wed, 08 Feb 2023 18:24:39 GMT",
      "comments": {}
    },
    {
      "version": "0.2.0",
      "tag": "@dossierhq/core-vitest_v0.2.0",
      "date": "Mon, 06 Feb 2023 21:59:52 GMT",
      "comments": {}
    },
    {
      "version": "0.1.53",
      "tag": "@dossierhq/core-vitest_v0.1.53",
      "date": "Mon, 06 Feb 2023 21:23:05 GMT",
      "comments": {
        "none": [
          {
            "comment": "Bump minor"
          }
        ]
      }
    },
    {
      "version": "0.1.52",
      "tag": "@dossierhq/core-vitest_v0.1.52",
      "date": "Tue, 31 Jan 2023 11:27:35 GMT",
      "comments": {}
    },
    {
      "version": "0.1.51",
      "tag": "@dossierhq/core-vitest_v0.1.51",
      "date": "Sun, 29 Jan 2023 18:08:56 GMT",
      "comments": {}
    },
    {
      "version": "0.1.50",
      "tag": "@dossierhq/core-vitest_v0.1.50",
      "date": "Sat, 28 Jan 2023 14:37:31 GMT",
      "comments": {}
    },
    {
      "version": "0.1.49",
      "tag": "@dossierhq/core-vitest_v0.1.49",
      "date": "Mon, 23 Jan 2023 09:38:38 GMT",
      "comments": {}
    },
    {
      "version": "0.1.48",
      "tag": "@dossierhq/core-vitest_v0.1.48",
      "date": "Sun, 22 Jan 2023 22:28:24 GMT",
      "comments": {}
    },
    {
      "version": "0.1.47",
      "tag": "@dossierhq/core-vitest_v0.1.47",
      "date": "Sun, 22 Jan 2023 22:01:52 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.46",
      "tag": "@dossierhq/core-vitest_v0.1.46",
      "date": "Tue, 17 Jan 2023 19:56:51 GMT",
      "comments": {}
    },
    {
      "version": "0.1.45",
      "tag": "@dossierhq/core-vitest_v0.1.45",
      "date": "Tue, 17 Jan 2023 17:55:40 GMT",
      "comments": {}
    },
    {
      "version": "0.1.44",
      "tag": "@dossierhq/core-vitest_v0.1.44",
      "date": "Tue, 17 Jan 2023 17:41:02 GMT",
      "comments": {}
    },
    {
      "version": "0.1.43",
      "tag": "@dossierhq/core-vitest_v0.1.43",
      "date": "Tue, 17 Jan 2023 10:34:29 GMT",
      "comments": {}
    },
    {
      "version": "0.1.42",
      "tag": "@dossierhq/core-vitest_v0.1.42",
      "date": "Tue, 17 Jan 2023 09:23:40 GMT",
      "comments": {}
    },
    {
      "version": "0.1.41",
      "tag": "@dossierhq/core-vitest_v0.1.41",
      "date": "Tue, 17 Jan 2023 08:54:04 GMT",
      "comments": {}
    },
    {
      "version": "0.1.40",
      "tag": "@dossierhq/core-vitest_v0.1.40",
      "date": "Mon, 16 Jan 2023 21:32:52 GMT",
      "comments": {}
    },
    {
      "version": "0.1.39",
      "tag": "@dossierhq/core-vitest_v0.1.39",
      "date": "Mon, 16 Jan 2023 21:07:04 GMT",
      "comments": {
        "none": [
          {
            "comment": "First release after @dossierhq/ move"
          }
        ]
      }
    },
    {
      "version": "0.1.38",
      "tag": "@dossierhq/core-vitest_v0.1.38",
      "date": "Mon, 02 Jan 2023 22:59:38 GMT",
      "comments": {}
    },
    {
      "version": "0.1.37",
      "tag": "@dossierhq/core-vitest_v0.1.37",
      "date": "Mon, 02 Jan 2023 17:39:42 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.36",
      "tag": "@dossierhq/core-vitest_v0.1.36",
      "date": "Mon, 12 Dec 2022 12:51:59 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.35",
      "tag": "@dossierhq/core-vitest_v0.1.35",
      "date": "Sat, 05 Nov 2022 21:33:57 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.34",
      "tag": "@dossierhq/core-vitest_v0.1.34",
      "date": "Tue, 25 Oct 2022 15:23:52 GMT",
      "comments": {}
    },
    {
      "version": "0.1.33",
      "tag": "@dossierhq/core-vitest_v0.1.33",
      "date": "Mon, 17 Oct 2022 15:41:41 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.32",
      "tag": "@dossierhq/core-vitest_v0.1.32",
      "date": "Fri, 30 Sep 2022 18:08:39 GMT",
      "comments": {}
    },
    {
      "version": "0.1.31",
      "tag": "@dossierhq/core-vitest_v0.1.31",
      "date": "Mon, 19 Sep 2022 21:05:31 GMT",
      "comments": {}
    },
    {
      "version": "0.1.30",
      "tag": "@dossierhq/core-vitest_v0.1.30",
      "date": "Mon, 19 Sep 2022 13:10:56 GMT",
      "comments": {}
    },
    {
      "version": "0.1.29",
      "tag": "@dossierhq/core-vitest_v0.1.29",
      "date": "Tue, 06 Sep 2022 09:47:25 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.28",
      "tag": "@dossierhq/core-vitest_v0.1.28",
      "date": "Tue, 30 Aug 2022 21:10:46 GMT",
      "comments": {}
    },
    {
      "version": "0.1.27",
      "tag": "@dossierhq/core-vitest_v0.1.27",
      "date": "Tue, 30 Aug 2022 14:05:13 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade heft"
          }
        ]
      }
    },
    {
      "version": "0.1.26",
      "tag": "@dossierhq/core-vitest_v0.1.26",
      "date": "Sun, 21 Aug 2022 22:26:51 GMT",
      "comments": {}
    },
    {
      "version": "0.1.25",
      "tag": "@dossierhq/core-vitest_v0.1.25",
      "date": "Sun, 21 Aug 2022 12:14:33 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies."
          }
        ]
      }
    },
    {
      "version": "0.1.24",
      "tag": "@dossierhq/core-vitest_v0.1.24",
      "date": "Mon, 13 Jun 2022 22:58:05 GMT",
      "comments": {}
    },
    {
      "version": "0.1.23",
      "tag": "@dossierhq/core-vitest_v0.1.23",
      "date": "Mon, 13 Jun 2022 21:34:43 GMT",
      "comments": {}
    },
    {
      "version": "0.1.22",
      "tag": "@dossierhq/core-vitest_v0.1.22",
      "date": "Mon, 13 Jun 2022 12:17:58 GMT",
      "comments": {}
    },
    {
      "version": "0.1.21",
      "tag": "@dossierhq/core-vitest_v0.1.21",
      "date": "Tue, 07 Jun 2022 21:00:46 GMT",
      "comments": {}
    },
    {
      "version": "0.1.20",
      "tag": "@dossierhq/core-vitest_v0.1.20",
      "date": "Mon, 06 Jun 2022 22:23:03 GMT",
      "comments": {}
    },
    {
      "version": "0.1.19",
      "tag": "@dossierhq/core-vitest_v0.1.19",
      "date": "Mon, 06 Jun 2022 15:26:55 GMT",
      "comments": {}
    },
    {
      "version": "0.1.18",
      "tag": "@dossierhq/core-vitest_v0.1.18",
      "date": "Mon, 06 Jun 2022 14:27:59 GMT",
      "comments": {}
    },
    {
      "version": "0.1.17",
      "tag": "@dossierhq/core-vitest_v0.1.17",
      "date": "Fri, 03 Jun 2022 19:47:15 GMT",
      "comments": {}
    },
    {
      "version": "0.1.16",
      "tag": "@dossierhq/core-vitest_v0.1.16",
      "date": "Thu, 02 Jun 2022 22:03:43 GMT",
      "comments": {}
    },
    {
      "version": "0.1.15",
      "tag": "@dossierhq/core-vitest_v0.1.15",
      "date": "Thu, 02 Jun 2022 21:27:42 GMT",
      "comments": {
        "none": [
          {
            "comment": "Upgrade dependencies"
          }
        ]
      }
    },
    {
      "version": "0.1.14",
      "tag": "@dossierhq/core-vitest_v0.1.14",
      "date": "Mon, 30 May 2022 18:59:49 GMT",
      "comments": {}
    },
    {
      "version": "0.1.13",
      "tag": "@dossierhq/core-vitest_v0.1.13",
      "date": "Mon, 30 May 2022 15:28:33 GMT",
      "comments": {}
    },
    {
      "version": "0.1.12",
      "tag": "@dossierhq/core-vitest_v0.1.12",
      "date": "Mon, 30 May 2022 11:14:25 GMT",
      "comments": {
        "none": [
          {
            "comment": "First release."
          }
        ]
      }
    }
  ]
}
