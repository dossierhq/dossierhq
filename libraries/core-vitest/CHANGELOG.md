# Change Log - @jonasb/datadata-core-vitest

This log was last generated on Mon, 30 May 2022 15:28:33 GMT and should not be manually modified.

## 0.1.13
Mon, 30 May 2022 15:28:33 GMT

_Version update only_

## 0.1.12
Mon, 30 May 2022 11:14:25 GMT

### Updates

- First release.

