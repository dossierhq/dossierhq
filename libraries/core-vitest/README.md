# `@dossierhq/core-vitest`

A test library for users of `@dossierhq/core` for usage with Vitest.

`@dossierhq/core-vitest` is part of [Dossier](https://www.dossierhq.dev/).

- [Changelog](./CHANGELOG.md)
