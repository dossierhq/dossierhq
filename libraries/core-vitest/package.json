{
  "name": "@dossierhq/core-vitest",
  "version": "0.7.21",
  "description": "A test library for users of @dossierhq/core for usage with Vitest",
  "keywords": [
    "dossier",
    "cms",
    "testing",
    "vitest"
  ],
  "homepage": "https://dossierhq.dev",
  "author": {
    "name": "Jonas Bengtsson"
  },
  "type": "module",
  "exports": {
    ".": "./lib/index.js"
  },
  "types": "./lib/index.d.ts",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsc",
    "build:watch": "tsc --watch",
    "lint": "npm run prettier:check && npm run eslint:check",
    "clean": "rm -rf lib/",
    "eslint:check": "eslint --max-warnings=0",
    "eslint:fix": "eslint --fix .",
    "prettier:check": "prettier --check .",
    "prettier:fix": "prettier --write ."
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/dossierhq/dossierhq.git",
    "directory": "libraries/core-vitest"
  },
  "dependencies": {
    "tslib": "~2.8.1"
  },
  "peerDependencies": {
    "@dossierhq/core": "workspace:*"
  },
  "devDependencies": {
    "@dossierhq/core": "workspace:*",
    "@dossierhq/eslint-config": "workspace:*",
    "@dossierhq/typescript-config": "workspace:*",
    "eslint": "~9.35.0",
    "typescript": "~5.8.3",
    "vitest": "~3.2.4"
  }
}
