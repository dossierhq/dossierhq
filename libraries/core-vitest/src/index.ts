export { expectErrorResult, expectOkResult, expectResultValue } from './ExpectResult.js';
export { createMockLogger } from './MockLogger.js';
