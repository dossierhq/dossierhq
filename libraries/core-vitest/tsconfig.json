{
  "extends": "@dossierhq/typescript-config/library.json",
  "compilerOptions": {
    "outDir": "lib"
  }
}
