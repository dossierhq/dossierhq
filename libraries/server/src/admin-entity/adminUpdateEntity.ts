import {
  AdminEntityStatus,
  notOk,
  ok,
  validateEntityInfoForUpdate,
  visitorPathToString,
  type AdminEntityMutationOptions,
  type AdminEntityUpdate,
  type AdminEntityUpdatePayload,
  type AdminSchema,
  type ErrorType,
  type PromiseResult,
  type PublishedSchema,
} from '@dossierhq/core';
import type { DatabaseAdapter } from '@dossierhq/database-adapter';
import { authVerifyAuthorizationKey } from '../Auth.js';
import type { AuthorizationAdapter } from '../AuthorizationAdapter.js';
import type { SessionContext } from '../Context.js';
import { encodeAdminEntity, resolveUpdateEntity } from '../EntityCodec.js';
import { randomNameGenerator } from './AdminEntityMutationUtils.js';
import { publishEntityAfterMutation } from './publishEntityAfterMutation.js';
import { updateUniqueIndexesForEntity } from './updateUniqueIndexesForEntity.js';

export async function adminUpdateEntity(
  adminSchema: AdminSchema,
  publishedSchema: PublishedSchema,
  authorizationAdapter: AuthorizationAdapter,
  databaseAdapter: DatabaseAdapter,
  context: SessionContext,
  entity: AdminEntityUpdate,
  options: AdminEntityMutationOptions | undefined
): PromiseResult<
  AdminEntityUpdatePayload,
  | typeof ErrorType.BadRequest
  | typeof ErrorType.NotFound
  | typeof ErrorType.NotAuthorized
  | typeof ErrorType.Generic
> {
  return await context.withTransaction(async (context) => {
    // get info of existing entity
    const entityInfoResult = await databaseAdapter.adminEntityUpdateGetEntityInfo(context, {
      id: entity.id,
    });
    if (entityInfoResult.isError()) return entityInfoResult;
    const {
      entityInternalId,
      name: previousName,
      authKey,
      resolvedAuthKey,
    } = entityInfoResult.value;

    // validate
    const validationIssue = validateEntityInfoForUpdate(
      ['entity'],
      { info: entityInfoResult.value },
      entity
    );
    if (validationIssue) {
      return notOk.BadRequest(
        `${visitorPathToString(validationIssue.path)}: ${validationIssue.message}`
      );
    }

    // auth key
    const authResult = await authVerifyAuthorizationKey(authorizationAdapter, context, {
      authKey,
      resolvedAuthKey,
    });
    if (authResult.isError()) return authResult;

    const resolvedResult = resolveUpdateEntity(adminSchema, entity, entityInfoResult.value);
    if (resolvedResult.isError()) return resolvedResult;
    const { changed, entity: updatedEntity, entitySpec } = resolvedResult.value;

    if (!changed) {
      const payload: AdminEntityUpdatePayload = { effect: 'none', entity: updatedEntity };
      if (options?.publish && updatedEntity.info.status !== AdminEntityStatus.published) {
        const publishResult = await publishEntityAfterMutation(
          adminSchema,
          publishedSchema,
          authorizationAdapter,
          databaseAdapter,
          context,
          {
            id: updatedEntity.id,
            version: updatedEntity.info.version,
          }
        );
        if (publishResult.isError()) return publishResult;
        payload.effect = 'published';
        updatedEntity.info.status = publishResult.value.status;
        updatedEntity.info.updatedAt = publishResult.value.updatedAt;
        updatedEntity.info.validPublished = true;
      }

      return ok(payload);
    }

    const encodeResult = await encodeAdminEntity(
      adminSchema,
      databaseAdapter,
      context,
      entitySpec,
      updatedEntity
    );
    if (encodeResult.isError()) return encodeResult;
    const { data, name } = encodeResult.value;

    const updateResult = await databaseAdapter.adminEntityUpdateEntity(
      context,
      randomNameGenerator,
      {
        entityInternalId,
        name: updatedEntity.info.name,
        changeName: name !== previousName,
        session: context.session,
        version: updatedEntity.info.version,
        status: updatedEntity.info.status,
        fieldValues: data,
      }
    );
    if (updateResult.isError()) return updateResult;

    let effect: AdminEntityUpdatePayload['effect'] = 'updated';
    updatedEntity.info.name = updateResult.value.name;
    updatedEntity.info.updatedAt = updateResult.value.updatedAt;

    const updateEntityIndexesResult = await databaseAdapter.adminEntityIndexesUpdateLatest(
      context,
      { entityInternalId },
      encodeResult.value.entityIndexes,
      false
    );
    if (updateEntityIndexesResult.isError()) return updateEntityIndexesResult;

    const uniqueIndexResult = await updateUniqueIndexesForEntity(
      databaseAdapter,
      context,
      { entityInternalId },
      false,
      encodeResult.value.uniqueIndexValues,
      null // TODO publishEntityAfterMutation is updating the values
    );
    if (uniqueIndexResult.isError()) return uniqueIndexResult;

    if (options?.publish) {
      const publishResult = await publishEntityAfterMutation(
        adminSchema,
        publishedSchema,
        authorizationAdapter,
        databaseAdapter,
        context,
        {
          id: updatedEntity.id,
          version: updatedEntity.info.version,
        }
      );
      if (publishResult.isError()) return publishResult;

      effect = 'updatedAndPublished';
      updatedEntity.info.status = publishResult.value.status;
      updatedEntity.info.updatedAt = publishResult.value.updatedAt;
      updatedEntity.info.validPublished = true;
    }

    return ok({ effect, entity: updatedEntity });
  });
}
