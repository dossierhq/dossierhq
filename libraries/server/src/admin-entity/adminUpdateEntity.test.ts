import { AdminEntityStatus, ErrorType, ok } from '@dossierhq/core';
import { expectErrorResult, expectResultValue } from '@dossierhq/core-vitest';
import { describe, expect, test } from 'vitest';
import {
  createMockAuthorizationAdapter,
  createMockDatabaseAdapter,
  createMockSessionContext,
  getDatabaseAdapterMockedCallsWithoutContextAndUnordered,
} from '../test/AdditionalTestUtils.js';
import { adminTestSchema, publishedTestSchema } from '../test/TestSchema.js';
import { adminUpdateEntity } from './adminUpdateEntity.js';

describe('Admin adminUpdateEntity', () => {
  test('Minimal', async () => {
    const databaseAdapter = createMockDatabaseAdapter();
    const authorizationAdapter = createMockAuthorizationAdapter();
    const context = createMockSessionContext({ databaseAdapter });
    const id = '123';
    const createdAt = new Date('2020-01-01T00:00:00.000Z');
    const now = new Date();

    authorizationAdapter.resolveAuthorizationKeys.mockReturnValueOnce(
      Promise.resolve(ok([{ authKey: 'none', resolvedAuthKey: 'none' }])),
    );
    databaseAdapter.adminEntityUpdateGetEntityInfo.mockReturnValueOnce(
      Promise.resolve(
        ok({
          entityInternalId: 'internal-123',
          type: 'TitleOnly',
          name: 'Old name',
          authKey: 'none',
          resolvedAuthKey: 'none',
          status: AdminEntityStatus.draft,
          valid: true,
          validPublished: null,
          version: 0,
          createdAt,
          updatedAt: createdAt,
          fieldValues: { title: 'Old title' },
        }),
      ),
    );
    databaseAdapter.adminEntityUpdateEntity.mockReturnValueOnce(
      Promise.resolve(ok({ name: 'Updated name#123456', version: 1, updatedAt: now })),
    );
    databaseAdapter.adminEntityIndexesUpdateLatest.mockReturnValueOnce(
      Promise.resolve(ok(undefined)),
    );
    databaseAdapter.adminEntityUniqueIndexGetValues.mockReturnValueOnce(Promise.resolve(ok([])));

    const result = await adminUpdateEntity(
      adminTestSchema,
      publishedTestSchema,
      authorizationAdapter,
      databaseAdapter,
      context,
      {
        id,
        info: { name: 'Updated name' },
        fields: { title: 'Updated title' },
      },
      undefined,
    );

    expectResultValue(result, {
      effect: 'updated',
      entity: {
        id: '123',
        info: {
          type: 'TitleOnly',
          name: 'Updated name#123456',
          authKey: 'none',
          version: 1,
          createdAt: createdAt,
          updatedAt: now,
          status: AdminEntityStatus.draft,
          valid: true,
          validPublished: null,
        },
        fields: {
          title: 'Updated title',
        },
      },
    });
    expect(
      getDatabaseAdapterMockedCallsWithoutContextAndUnordered(databaseAdapter),
    ).toMatchSnapshot();
  });

  test('Error: No change to invalid entity', async () => {
    const databaseAdapter = createMockDatabaseAdapter();
    const authorizationAdapter = createMockAuthorizationAdapter();
    const context = createMockSessionContext({ databaseAdapter });
    const id = '123';
    const createdAt = new Date('2020-01-01T00:00:00.000Z');

    authorizationAdapter.resolveAuthorizationKeys.mockReturnValueOnce(
      Promise.resolve(ok([{ authKey: 'none', resolvedAuthKey: 'none' }])),
    );
    databaseAdapter.adminEntityUpdateGetEntityInfo.mockReturnValueOnce(
      Promise.resolve(
        ok({
          entityInternalId: 'internal-123',
          type: 'TitleOnly',
          name: 'Old name',
          authKey: 'none',
          resolvedAuthKey: 'none',
          status: AdminEntityStatus.draft,
          valid: false,
          validPublished: null,
          version: 0,
          createdAt,
          updatedAt: createdAt,
          fieldValues: { title: 'Old title' },
        }),
      ),
    );

    const result = await adminUpdateEntity(
      adminTestSchema,
      publishedTestSchema,
      authorizationAdapter,
      databaseAdapter,
      context,
      { id, fields: {} },
      undefined,
    );

    expectErrorResult(result, ErrorType.BadRequest, 'No change to entity that is already invalid');
    expect(
      getDatabaseAdapterMockedCallsWithoutContextAndUnordered(databaseAdapter),
    ).toMatchSnapshot();
  });
});
