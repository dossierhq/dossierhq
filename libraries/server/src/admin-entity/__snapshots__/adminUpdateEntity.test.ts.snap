// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Admin adminUpdateEntity > Error: No change to invalid entity 1`] = `
[
  [
    "adminEntityUpdateGetEntityInfo",
    {
      "id": "123",
    },
  ],
  [
    "withRootTransaction",
    [Function],
    [Function],
  ],
]
`;

exports[`Admin adminUpdateEntity > Minimal 1`] = `
[
  [
    "adminEntityIndexesUpdateLatest",
    {
      "entityInternalId": "internal-123",
    },
    {
      "componentTypes": [],
      "fullTextSearchText": "Updated title",
      "locations": [],
      "referenceIds": [],
    },
    false,
  ],
  [
    "adminEntityUniqueIndexGetValues",
    {
      "entityInternalId": "internal-123",
    },
  ],
  [
    "adminEntityUpdateEntity",
    [Function],
    {
      "changeName": true,
      "encodeVersion": 1,
      "entityInternalId": "internal-123",
      "fields": {
        "title": "Updated title",
      },
      "name": "Updated name",
      "publish": false,
      "schemaVersion": 1,
      "session": {
        "subjectId": "subject-id",
        "subjectInternalId": 123,
      },
      "status": "draft",
      "type": "TitleOnly",
      "version": 2,
    },
    null,
  ],
  [
    "adminEntityUpdateGetEntityInfo",
    {
      "id": "123",
    },
  ],
  [
    "withRootTransaction",
    [Function],
    [Function],
  ],
]
`;
